Programs by ProkhorovMax (13531/3)

Prokhorovmax opened the pull request.

Commit:
Тривиальная 1

Commit:
Последняя тривиальная и первая и последняя простые

Commit:
Простая один урок 2, задание 1

Commit:
l2 t1 easy 1

Commit:
l2 t1 easy 1

Commit:
l2 t1 easy 2

Commit:
l2 t1 easy 3

Prokhorovmax committed:
l2 t1 e1

Commit:
l2 t1 normal

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
l2 t2 normal 1(circle inside)

Commit:
l2 t2 normal 2(brick passes)

Prokhorovmax committed:
Исправлено
l2 t1 easy 2,3; normal 1;
l2 t2 normal 1,2

Prokhorovmax committed:
Исправлено
l1 third digit

Prokhorovmax committed:
l3:
trivial
easy 1 (fib)
easy 2 (lcm)
easy 4 (maxDivisor)
easy 5 (isCoPrime)
easy 8 (cos)

Prokhorovmax committed:
l3:
trivial (Исправление)

ice-phoenix commented:
recheck all

ice-phoenix commented:
recheck all

ice-phoenix commented on src/lesson1/task1/Simple.kt:
```diff
@@ -51,7 +51,12 @@ fun main(args: Array<String>) {
  */
-fun seconds(hours: Int, minutes: Int, seconds: Int): Int = TODO()
+fun seconds(hours: Int, minutes: Int, seconds: Int): Int {
+    val a = hours*3600
```
Переменным лучше давать осмысленные имена, соответствующие их назначению в коде

Prokhorovmax committed:
l3: trivial
HELP! Не могу понять, почему модуль не решает проблему в том случае, когда на вход подается отрицательное число. Подскажите, что не так :/

ice-phoenix commented on src/lesson1/task1/Simple.kt:
```diff
@@ -76,15 +81,21 @@ fun angleInRadian(grad: Int, min: Int, sec: Int): Double = TODO()
  */
-fun trackLength(x1: Double, y1: Double, x2: Double, y2: Double): Double = TODO()
+fun trackLength(x1: Double, y1: Double, x2: Double, y2: Double): Double {
+    val dlina = sqrt (sqr(x2-x1)+sqr(y2-y1))
```
Использование транслитерации в имени переменной --- **очень плохой** стиль

Prokhorovmax replied:
Я давно это понял. Извините, просто забыл исправить старые ошибки.

ice-phoenix commented on src/lesson1/task1/Simple.kt:
```diff
@@ -110,4 +121,12 @@ fun accountInThreeYears(initial: Int, percent: Int): Double = TODO()
  */
-fun numberRevert(number: Int): Int = TODO()
+fun numberRevert(number: Int): Int {
+    val a = number%10
```
Переменным лучше давать осмысленные имена, соответствующие их назначению в коде

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -33,7 +33,13 @@ fun minBiRoot(a: Double, b: Double, c: Double): Double {
-fun ageDescription(age: Int): String = TODO()
+fun ageDescription(age: Int): String {
+   return when {
+       (age % 10 == 1)&&(age%100 != 11) -> "$age год"
```
Отформатируйте код при помощи IntelliJ IDEA и в дальнейшем постарайтесь придерживаться данного стиля кодирования

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +72,14 @@ fun timeForHalfWay(t1: Double, v1: Double,
                        rookX1: Int, rookY1: Int,
-                       rookX2: Int, rookY2: Int): Int = TODO()
+                       rookX2: Int, rookY2: Int): Int {
+    if (((kingX == rookX1) || (kingY == rookY1)) && ((kingX == rookX2) || (kingY == rookY2))) return 3
```
Выражения, проверяющие факт атаки короля той или иной фигурой, следует вынести в отдельные переменные

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -89,4 +112,11 @@ fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
-fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int = TODO()
\ No newline at end of file
+fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int {
+    return  if ((c>=a) && (d<=b)) d-c
```
Данное выражение лучше оформить как `when`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -1,6 +1,8 @@
 @file:Suppress("UNUSED_PARAMETER")
 package lesson3.task1
 
+import lesson2.task1.timeForHalfWay
```
Зачем в `lesson3` импорт функции из `lesson2`?

Prokhorovmax replied:
Без понятия, откуда там эта строчка и зачем.

ice-phoenix replied:
Как разработчик, вы отвечаете за весь код и все изменения, сделанный вами. Вариант "без понятия, откуда что-то где-то", к сожалению, не подходит ни для нас во время учебы, ни для вашего начальника во время работы. В будущем просматривайте изменения перед комитом и убеждайтесь в том, что там нет ничего лишнего.

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,23 +59,48 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+fun digitNumber(n: Int): Int {
+    if (n == 0) return 1
+    else {
+    var number = n
```
Форматирование кода

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,23 +59,48 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+            res = i
+            break
+        }
+        else continue
```
Данная конструкция является излишней

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -87,7 +114,13 @@ fun minDivisor(n: Int): Int = TODO()
+fun maxDivisor(n: Int): Int {
+    var res = 0
+    for (i in 1..n/2) {
+        if (n%i == 0) res = i
```
Зачем при поиске максимального делителя идти от минимального делителя к максимальному?

Prokhorovmax replied:
@ice-phoenix, я не могу реализовать программу, когда надо перебирать значения от большего к меньшему(любую программу). У меня выделяется красным "downto", поэтому я не понимаю, что делать в данном случае и еще не могу решить пару других задач, где это нужно. Был бы признателен, если бы вы объяснили, что не так. Спасибо.

ice-phoenix replied:
Можно попробовать подойти к преподавателю на занятиях и спросить, что не так

mglukhikh replied:
Раз уж заметил: нужно писать `downTo`. Это ошибка в учебном пособии, которая сейчас уже исправлена.

ice-phoenix reviewed (changes requested):
Исправляйте **все** предыдущие замечания к вашему коду перед подачей следующих PR

Prokhorovmax committed:
Внесены исправления в трех уроках

Prokhorovmax committed:
l3 normal 1
Внесены исправления

Prokhorovmax committed:
l3 hard 1

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -157,7 +226,15 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
+    var i: Int = 1
+    var str: String = ""
+    while (str.length < n) {
+        str += (i * i).toString()
```
Подумайте, строка какой длины потребуется для того, чтобы посчитать `squareSequenceDigit(1000000)`. Можно ли обойтись без хранения значений промежуточных квадратов?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +163,43 @@ fun sin(x: Double, eps: Double): Double = TODO()
+    var number = n
+    var result = 0.0
+    for (i in 1..length) {
+        result += number % 10 * Math.pow(10.toDouble(), length - i.toDouble())
```
Подумайте, как можно организовать ваш код так, чтобы не было необходимо каждый раз возводить 10 в требуемую степень с самого начала

Commit:
l3 hard 1 (squareSequenceDigit)

belyaev-mikhail commented:
recheck assignment

Commit:
l4 easy 1 (abs)

Commit:
l4 normal 1 (center)

ice-phoenix commented:
recheck all

ice-phoenix commented:
Просьба исправить предыдущие замечания перед тем, как продолжать работу над следующими модулями

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -104,7 +104,17 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
+    else {
+        var sum = 0.0
+        for (i in 0..length-1) {
+            sum += v[i] * v[i]
```
Если вам не нужен явный индекс, то лучше воспользоваться циклом в `foreach` стиле.

```
for (e in list) {
    foo(e)
}
```

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -119,7 +129,22 @@ fun mean(list: List<Double>): Double = TODO()
+        for (i in 0..length - 1) {
+            average += list[i]
+        }
+        average /= length
```
Вместо всего этого можно просто `list.sum() / list.size`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -157,7 +226,20 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
+        sum += digitNumber(square)
+    }
+    count = sum - n
+    number = square.toDouble()/Math.pow(10.0, (count).toDouble())
```
Вычисления в `Double` могут выполняться с недостаточной точностью. Нельзя ли организовать код так, чтобы все вычислять в `Int`?

ice-phoenix reviewed (changes requested):
Кроме этих замечаний, исправьте предыдущие замечания

Prokhorovmax committed:
l3 easy (cos) - попытка исправить

Prokhorovmax committed:
l3 normal (revert) - исправлено

Prokhorovmax committed:
l4 easy (abs) - исправлено

Prokhorovmax committed:
l4 normal (center) - исправлено

Prokhorovmax committed:
l3 hard (squareSequenceDigit) - исправлено

ice-phoenix commented:
recheck all

ice-phoenix commented:
recheck all

Prokhorovmax committed:
l3 easy (cos) - Исправлено

Prokhorovmax committed:
l4 normal (decimal)

Prokhorovmax committed:
l4 hard (convertToString)

ice-phoenix commented on src/lesson1/task1/Simple.kt:
```diff
@@ -51,7 +51,11 @@ fun main(args: Array<String>) {
  */
-fun seconds(hours: Int, minutes: Int, seconds: Int): Int = TODO()
+fun seconds(hours: Int, minutes: Int, seconds: Int): Int {
+    val hoursInSeconds = hours*3600
```
Форматирование кода должно быть единообразным. Если вы разделяете операции пробелами, то делайте это везде. Если вы пишете операции вместе с операндами, то тоже делайте это везде. У вас же стиль плавает туда-сюда, что не есть очень хорошо.

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -44,7 +51,17 @@ fun ageDescription(age: Int): String = TODO()
+    val s3 = v3 * t3
+    val s = s1 + s2 + s3
+    val half = s / 2
+    val res = if (half <= s1) half / v1
```
Если вы используете `if` как выражение, то форматирование должно выглядеть как

```
val a = if (...) ...
        else if (...) ...
        else ...
```

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +73,18 @@ fun timeForHalfWay(t1: Double, v1: Double,
                        rookX1: Int, rookY1: Int,
-                       rookX2: Int, rookY2: Int): Int = TODO()
+                       rookX2: Int, rookY2: Int): Int {
+    var first = 0
```
- Почему `first` и `second` не типа `Boolean`?
- Почему их значения не инициализируются сразу при объявлении?

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +73,18 @@ fun timeForHalfWay(t1: Double, v1: Double,
+    var second = 0
+    if ((kingX == rookX1) || (kingY == rookY1)) first = 1
+    if ((kingX == rookX2) || (kingY == rookY2)) second = 1
+    if ((first == 1) && (second == 1)) return 3
```
Данное выражение лучше оформить как `when`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,23 +59,47 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+    val max = Math.max(m, n)
+    for (i in max..m * n) {
+        if ((i % m == 0) && (i % n == 0)) {
+            res = i
```
Можно обойтись без переменной `res`, если сразу возвращать результат работы функции

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -87,7 +113,16 @@ fun minDivisor(n: Int): Int = TODO()
+    var res = 0
+    for (i in n / 2 downTo 1) {
+        if (n % i == 0) {
+            res = i
```
Можно обойтись без переменной res, если сразу возвращать результат работы функции

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,7 +131,12 @@ fun maxDivisor(n: Int): Int = TODO()
-fun isCoPrime(m: Int, n: Int): Boolean = TODO()
+fun isCoPrime(m: Int, n: Int): Boolean {
+    for (i in 2..Math.max(m, n)) {
+        if ((m % i == 0) && (n % i == 0)) return false
```
Можно ли данную функцию выразить через другие функции данного модуля?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +163,44 @@ fun sin(x: Double, eps: Double): Double = TODO()
+    var res = 1.0
+    var a = 2.0
+    var b = 2
+    var sign = 1 // Значение будет определять, с каким знаком брать очередной член ряда.
```
Почему у данной переменной тип не `Boolean`?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +205,30 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+        }
+        for (i in 0..length - 1) {
+            if (numerals[i].toInt() > 9) {
+                numerals[i] = alphabet[(numerals[i].toInt() - 10)].toString()
```
Почему нельзя сразу записывать правильное представление очередной цифры в итоговый список?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -193,7 +237,14 @@ fun convertToString(n: Int, base: Int): String = TODO()
+    val length = digits.size
+    var res = 0
+    for (i in 0..length - 1) {
+        res += digits[i] * Math.pow(base.toDouble(), (length - i - 1).toDouble()).toInt()
```
Лучше было пару уроков назад написать `pow`, принимающий `Int`

ice-phoenix commented:
recheck all

ice-phoenix commented:
Исправляйте замечания (см. вкладку `Files changed`) и продолжаем работать дальше =)

Prokhorovmax committed:
l1 - внесены исправления

Prokhorovmax committed:
l2 t1 - внесены исправления

Prokhorovmax committed:
l2 t2 - внесены исправления

Prokhorovmax committed:
l3 - внесены исправления

ice-phoenix commented:
recheck all

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -44,7 +51,16 @@ fun ageDescription(age: Int): String = TODO()
+    val s3 = v3 * t3
+    val s = s1 + s2 + s3
+    val half = s / 2
+    return if (half <= s1) half / v1
```
Данное выражение лучше оформить как `when`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +72,20 @@ fun timeForHalfWay(t1: Double, v1: Double,
                        rookX1: Int, rookY1: Int,
-                       rookX2: Int, rookY2: Int): Int = TODO()
+                       rookX2: Int, rookY2: Int): Int {
+    var first : Boolean = false
```
- Система типов Котлина выведет тип переменной автоматически

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +72,20 @@ fun timeForHalfWay(t1: Double, v1: Double,
+                       rookX2: Int, rookY2: Int): Int {
+    var first : Boolean = false
+    var second : Boolean = false
+    if ((kingX == rookX1) || (kingY == rookY1)) first = true
```
Если присвоить переменной значение сразу при объявлении, то можно будет сделать ее `val`

Prokhorovmax replied:
Я не очень понимаю, что мы выиграем от этого, да и вдобавок компилятор не позволяет мне так сделать.

Prokhorovmax replied:
Мне подсказали, что имелось в виду, исправил.

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -1,6 +1,10 @@
+
 package lesson3.task1
 
+import java.lang.reflect.Executable
```
Зачем вам этот импорт?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,23 +61,45 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  */
-fun fib(n: Int): Int = TODO()
+fun fib(n: Int): Int =
+        if ((n == 1) || (n == 2)) 1
```
Лишние скобки

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,23 +61,45 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+            return i
+        }
+    }
+    throw Exception()
```
Что?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,23 +61,45 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+    val max = Math.max(m, n)
+    for (i in max..m * n) {
+        if ((i % m == 0) && (i % n == 0)) {
+            return i
```
Почитайте про более быстрые алгоритмы поиска НОК

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -87,7 +113,14 @@ fun minDivisor(n: Int): Int = TODO()
+            return i
+        }
+    }
+    throw Exception()
```
Зачем?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,7 +129,12 @@ fun maxDivisor(n: Int): Int = TODO()
-fun isCoPrime(m: Int, n: Int): Boolean = TODO()
+fun isCoPrime(m: Int, n: Int): Boolean {
+    for (i in 2..Math.min(m, n)) {
+        if ((m % i == 0) && (n % i == 0)) return false
```
Нельзя ли эту функцию выразить через другие функции, разработанные вами в этом модуле?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +161,44 @@ fun sin(x: Double, eps: Double): Double = TODO()
+            res += Math.pow(x1, a) / factorial(b)
+            a += 2
+            b += 2
+            sign = true
```
Лучше сделать `sign = !sign` и вынести это из `if`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +161,44 @@ fun sin(x: Double, eps: Double): Double = TODO()
+            b += 2
+            sign = true
+        }
+    } while (Math.abs(Math.pow(x1, a) / factorial(b)) >= eps)
```
Повторно используемое выражение (тем более, с факториалом) лучше не считать два раза

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -157,7 +225,19 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
+        sum += digitNumber(square)
+    }
+    count = sum - n
+    return ((square / (Math.pow(10.0, (count).toDouble()))) % 10).toInt()
```
Напишите `pow`, который работает с целыми числами

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -104,7 +105,16 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
  */
-fun abs(v: List<Double>): Double = TODO()
+fun abs(v: List<Double>): Double {
+    if (v.size == 0) return 0.0
```
Подумайте, нужна ли эта проверка? Сколько раз выполнится тело цикла ниже в случае, если список пуст?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +205,30 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+fun convertToString(n: Int, base: Int): String {
+    val alphabet = listOf<Char>('a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z')
+    var number = n
+    val revertedNumerals = mutableListOf<Int>()
```
Есть функция `reverse() / reversed()`, которая вроде бы делает то, что вам надо, но несколько попроще

Prokhorovmax replied:
К сожалению, эта функция не позволяет мне далее использовать индексы. Но, возможно что-то я делаю не так; спрошу у преподавателя на паре.

Prokhorovmax replied:
Все узнал, получилось.

ice-phoenix reviewed (changes requested):
Исправляем замечания и работаем дальше

Prokhorovmax committed:
l2 t1 / l3 - внесены исправления

Prokhorovmax committed:
l3:
+ pow
+ gcd
Внесены исправления

Prokhorovmax committed:
l4 - внесены исправления

Prokhorovmax committed:
l2 t1 (whichRookThreatens) - исправлено
l4 (convertToString) - исправлено

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -35,7 +35,9 @@ fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean = TODO()
 fun circleInside(x1: Double, y1: Double, r1: Double,
-                 x2: Double, y2: Double, r2: Double): Boolean = TODO()
+                 x2: Double, y2: Double, r2: Double): Boolean =
+    (Math.sqrt(sqr(x2 - x1) + sqr(y2 - y1))) <= (r2 - r1)
```
Лишние скобки вокруг законченного выражения `Math.sqrt(...)`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,23 +59,52 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  */
-fun lcm(m: Int, n: Int): Int = TODO()
+fun lcm(m: Int, n: Int): Int =
+        Math.abs(m * n) / gcd(m, n)
```
Зачем вызов модуля?

ice-phoenix replied:
Данное замечание все еще в силе

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,7 +134,8 @@ fun maxDivisor(n: Int): Int = TODO()
  */
-fun isCoPrime(m: Int, n: Int): Boolean = TODO()
+fun isCoPrime(m: Int, n: Int): Boolean =
+        gcd(m, n) == 1
```
:+1:

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +204,31 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
  */
-fun convertToString(n: Int, base: Int): String = TODO()
+fun convertToString(n: Int, base: Int): String {
+    val alphabet = listOf<Char>('a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z')
```
В принципе, это можно заменить на строку `"abc...xyz"`

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +204,31 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+            if (revertedNumerals[i].toInt() > 9) {
+                numerals.add(alphabet[(revertedNumerals[i] - 10)].toString())
+            }
+            else numerals.add(revertedNumerals[i].toString())
```
Весь этот цикл лучше заменить на вызов функции `map` над `revertedNumerals`

Prokhorovmax replied:
Начал делать с помощью map, не все получалось, но в итоге сделал без map и существенно проще и короче.

ice-phoenix reviewed (changes requested):
Спасибо за исправления, все очень и очень здорово! :sunglasses:

Prokhorovmax committed:
l2 t2 - внесены исправления
convertToString (l4) - изменена

Prokhorovmax committed:
l5 - hard (bestHighJump)

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +204,21 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+        }
+        numerals.add(alphabet[number % base])
+        numerals.reverse()
+        return numerals.joinToString("")
```
Лучше сразу `numerals.reversed().joinToString(...)`

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +106,18 @@ fun bestLongJump(jumps: String): Int = TODO()
+    var maxHigh = -1
+    for (i in 1..parts.size - 1 step 2) {
+        for (e in parts[i]) {
+            if ((e == '+') && (parts[i - 1].toInt() > maxHigh)) {
```
Вместо цикла можно `parts[i].contains(...) && ...`

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +106,18 @@ fun bestLongJump(jumps: String): Int = TODO()
+    for (i in 1..parts.size - 1 step 2) {
+        for (e in parts[i]) {
+            if ((e == '+') && (parts[i - 1].toInt() > maxHigh)) {
+                maxHigh = parts[i - 1].toInt()
```
Повторно используемые подвыражения лучше вынести в отдельные переменные с говорящими именами

Prokhorovmax committed:
l3 (lcm) / l4 (convertToString) / l5 (bestHighJump) - исправлены

Prokhorovmax committed:
bestHighJump (l5) - исправлена

ice-phoenix commented:
Все вообще отлично. Закрываю данный PR, чтобы облегчить жизнь GitHub'у, не забудьте создать новый PR после того, как решите следующие задачи.
